spec/function.dd

```diff
@@ -2884,10 +2884,60 @@ $(H3 $(LNAME2 safe-functions, Safe Functions))
         so they can be corrected.
         )
 
+$(H4 Safe External Functions)
+
+        $(P External functions don't have a function body visible to the compiler:
+        ---
+        @safe extern (C) void play();
+        ---
+        and so safety cannot be verified automatically.
+        )
+
+        $(BEST_PRACTICE Explicitly set an attribute for external functions rather
+        than relying on default settings.)
+
 $(H3 $(LNAME2 trusted-functions, Trusted Functions))
 
         $(P Trusted functions are marked with the $(CODE @trusted) attribute.)
 
+        $(P Trusted functions act as an interface between safe and system functions.
+        A trusted function provides a safe interface. A $(I Safe Interface) is one that:
+
+        $(UL
+
+        $(LI Does not invalidate pointers that are passed to it. For example,
+        ---
+        @system void free(void* p);
+        ...
+        free(p);
+        *p = 3;  // p is invalid
+        ---
+        is not a safe interface because `free(p)` means `p` is no longer a valid
+        pointer when it returns. `free` can only be `@system`.)
+
+        $(LI Does not attempt to access memory beyond the bounds of the memory object
+        a parameter points to. For example:
+        ---
+        @system size_t strlen(char* p);
+        ---
+        is not a safe interface because the implementation of `strlen` advances `p` until
+        it finds a `null` character, and this can be arbitrarily beyond the `char` that `p`
+        purportedly points to. `strlen` can only be `@system`. Any function that traverses
+        a C string passed as an argument can only be `@system`.
+        )
+
+        $(LI Has array bounds passed as a separate parameter rather than as part of the pointer
+        type. For example:
+        ---
+        @system void* memcpy(void* dst, void* src, size_t nbytes);
+        ---
+        is not a safe interface because the extent of `dst` and `src` is set by `nbytes`, and
+        so cannot be verified by the compiler. Such functions can only be `@system`.
+        )
+
+        )
+        )
```

Comment:
I suggest merging this whole new section into the one below it ("Trusted functions are guaranteed to [...]"). The existing paragraph already says that `@trusted` functions must be safe. Maybe add the phrase "Safe Interface" there. And there is already a block of examples of things that can't be `@trusted`. Maybe add your examples there.

Reply:
It's better in the trusted section because many people misunderstand what trusted is for.

Reply:
> It's better in the trusted section

I'm not saying to move it out of the trusted section. Just merge it better with the rest of it. The text is somewhat repeating itself right now.

Instead of this structure:

* your definition of `@trusted` with "safe interface"
* examples of things that can't be `@trusted`
* existing definition of `@trusted`
* more examples of things that can't be `@trusted`

Do this:

* Expand existing definition of `@trusted` to include "safe interface".
* Add new examples next to existing examples.